<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Parse Biocompute Object From JSON File to R Object — parse_bco • biocompute</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Parse Biocompute Object From JSON File to R Object — parse_bco" />
<meta property="og:description" content="Parse Biocompute Object From JSON File to R Object" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">biocompute</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/case-study.html">Authoring Biocompute Objects with R: A Case Study</a>
    </li>
    <li>
      <a href="../articles/intro.html">Create and Manipulate BioCompute Objects with R</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/sbg/biocompute/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Parse Biocompute Object From JSON File to R Object</h1>
    <small class="dont-index">Source: <a href='https://github.com/sbg/biocompute/blob/master/R/parse-bco.R'><code>R/parse-bco.R</code></a></small>
    <div class="hidden name"><code>parse_bco.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Parse Biocompute Object From JSON File to R Object</p>
    </div>

    <pre class="usage"><span class='fu'>parse_bco</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>...</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>BioCompute Object .json file</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional parameters for <code><a href='https://rdrr.io/pkg/jsonlite/man/fromJSON.html'>fromJSON</a></code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list of class <code>bco</code></p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='va'>bco</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempfile</a></span><span class='op'>(</span>fileext <span class='op'>=</span> <span class='st'>".json"</span><span class='op'>)</span>
<span class='va'>bco</span> <span class='op'>&lt;-</span> <span class='fu'><a href='generate_example.html'>generate_example</a></span><span class='op'>(</span><span class='st'>"HCV1a"</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='convert_json.html'>convert_json</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='export_json.html'>export_json</a></span><span class='op'>(</span><span class='va'>bco</span><span class='op'>)</span>
<span class='va'>bco</span> <span class='op'>%&gt;%</span> <span class='fu'>parse_bco</span><span class='op'>(</span><span class='op'>)</span>
</div><div class='output co'>#&gt; $spec_version
#&gt; [1] "https://w3id.org/biocompute/1.4.2/"
#&gt; 
#&gt; $object_id
#&gt; [1] "https://biocompute.sbgenomics.com/bco/abdc4e88-405b-48d2-bbe4-7b2a02b8ff68"
#&gt; 
#&gt; $etag
#&gt; [1] "d18deb41a97a3108e743231af5fec0055dac321b945a569ca58bf33503ec526b"
#&gt; 
#&gt; $provenance_domain
#&gt; $provenance_domain$name
#&gt; [1] "HCV1a ledipasvir resistance SNP detection"
#&gt; 
#&gt; $provenance_domain$version
#&gt; [1] "1.0.0"
#&gt; 
#&gt; $provenance_domain$review
#&gt;     status
#&gt; 1 approved
#&gt; 2 approved
#&gt;                                                           reviewer_comment
#&gt; 1 Approved by [company name] staff. Waiting for approval from FDA Reviewer
#&gt; 2                                               The revised BCO looks fine
#&gt;         date
#&gt; 1 1510507848
#&gt; 2 1513110648
#&gt;                                                                                                       reviewer
#&gt; 1 Jane Doe, Seven Bridges Genomics, example@sevenbridges.com, curatedBy, https://orcid.org/0000-0000-0000-0000
#&gt; 2                                  John Doe, U.S. Food and Drug Administration, example@fda.gov, curatedBy, NA
#&gt; 
#&gt; $provenance_domain$derived_from
#&gt; [1] "https://github.com/biocompute-objects/BCO_Specification/blob/1.2.1-beta/HCV1a.json"
#&gt; 
#&gt; $provenance_domain$obsolete_after
#&gt; [1] "2018-11-12T12:30:48-0500"
#&gt; 
#&gt; $provenance_domain$embargo
#&gt; [1] "2017-10-12T12:30:48-0500" "2017-11-12T12:30:48-0500"
#&gt; 
#&gt; $provenance_domain$created
#&gt; [1] "2017-01-20T09:40:17-0500"
#&gt; 
#&gt; $provenance_domain$modified
#&gt; [1] "2019-05-10T09:40:17-0500"
#&gt; 
#&gt; $provenance_domain$contributors
#&gt;       name                       affiliation                    email
#&gt; 1 Jane Doe            Seven Bridges Genomics example@sevenbridges.com
#&gt; 2 John Doe U.S. Food and Drug Administration          example@fda.gov
#&gt;           contribution                                 orcid
#&gt; 1 createdBy, curatedBy https://orcid.org/0000-0000-0000-0000
#&gt; 2           authoredBy                                    NA
#&gt; 
#&gt; $provenance_domain$license
#&gt; [1] "https://creativecommons.org/licenses/by/4.0/"
#&gt; 
#&gt; 
#&gt; $usability_domain
#&gt; [1] "Identify baseline single nucleotide polymorphisms (SNPs)[SO:0000694], (insertions)[SO:0000667], and (deletions)[SO:0000045] that correlate with reduced (ledipasvir)[pubchem.compound:67505836] antiviral drug efficacy in (Hepatitis C virus subtype 1)[taxonomy:31646]"
#&gt; [2] "Identify treatment emergent amino acid (substitutions)[SO:1000002] that correlate with antiviral drug treatment failure"                                                                                                                                                 
#&gt; [3] "Determine whether the treatment emergent amino acid (substitutions)[SO:1000002] identified correlate with treatment failure involving other drugs against the same virus"                                                                                                
#&gt; 
#&gt; $extension_domain
#&gt; $extension_domain$fhir_extension
#&gt; $extension_domain$fhir_extension$fhir_endpoint
#&gt; [1] "https://fhirtest.uhn.ca/baseDstu3"
#&gt; 
#&gt; $extension_domain$fhir_extension$fhir_version
#&gt; [1] "3"
#&gt; 
#&gt; $extension_domain$fhir_extension$fhir_resources
#&gt; $extension_domain$fhir_extension$fhir_resources[[1]]
#&gt;   fhir_id fhir_resource
#&gt; 1   21376      Sequence
#&gt; 
#&gt; $extension_domain$fhir_extension$fhir_resources[[2]]
#&gt;   fhir_id    fhir_resource
#&gt; 1 6288583 DiagnosticReport
#&gt; 
#&gt; $extension_domain$fhir_extension$fhir_resources[[3]]
#&gt;   fhir_id    fhir_resource
#&gt; 1   25544 ProcedureRequest
#&gt; 
#&gt; $extension_domain$fhir_extension$fhir_resources[[4]]
#&gt;   fhir_id fhir_resource
#&gt; 1   92440   Observation
#&gt; 
#&gt; $extension_domain$fhir_extension$fhir_resources[[5]]
#&gt;   fhir_id       fhir_resource
#&gt; 1 4588936 FamilyMemberHistory
#&gt; 
#&gt; 
#&gt; 
#&gt; $extension_domain$scm_extension
#&gt; $extension_domain$scm_extension$scm_repository
#&gt; [1] "https://github.com/example/repo1"
#&gt; 
#&gt; $extension_domain$scm_extension$scm_type
#&gt; [1] "git"
#&gt; 
#&gt; $extension_domain$scm_extension$scm_commit
#&gt; [1] "c9ffea0b60fa3bcf8e138af7c99ca141a6b8fb21"
#&gt; 
#&gt; $extension_domain$scm_extension$scm_path
#&gt; [1] "workflow/hive-viral-mutation-detection.cwl"
#&gt; 
#&gt; $extension_domain$scm_extension$scm_preview
#&gt; [1] "https://github.com/example/repo1/blob/c9ffea0b60fa3bcf8e138af7c99ca141a6b8fb21/workflow/hive-viral-mutation-detection.cwl"
#&gt; 
#&gt; 
#&gt; 
#&gt; $description_domain
#&gt; $description_domain$keywords
#&gt; [1] "HCV1a"                    "Ledipasvir"              
#&gt; [3] "antiviral resistance"     "SNP"                     
#&gt; [5] "amino acid substitutions"
#&gt; 
#&gt; $description_domain$xref
#&gt;          namespace              name
#&gt; 1 pubchem.compound  PubChem-compound
#&gt; 2           pubmed            PubMed
#&gt; 3               so Sequence Ontology
#&gt; 4         taxonomy          Taxonomy
#&gt;                                             ids              access_time
#&gt; 1                                      67505836 2017-01-20T09:40:17-0500
#&gt; 2                                      26508693 2017-01-21T09:40:17-0500
#&gt; 3 SO:000002, SO:0000694, SO:0000667, SO:0000045 2017-01-22T09:40:17-0500
#&gt; 4                                         31646 2017-01-23T09:40:17-0500
#&gt; 
#&gt; $description_domain$platform
#&gt; [1] "Seven Bridges Platform"
#&gt; 
#&gt; $description_domain$pipeline_steps
#&gt;   step_number         name                               description version
#&gt; 1           1 HIVE-hexagon Alignment of reads to a set of references     1.3
#&gt;                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         prerequisite
#&gt; 1 Hepatitis C virus genotype 1, Hepatitis C virus type 1b complete genome, Hepatitis C virus (isolate JFH-1) genomic RNA, Hepatitis C virus clone J8CF, complete genome, Hepatitis C virus S52 polyprotein gene, https://www.ncbi.nlm.nih.gov/nuccore/22129792, https://www.ncbi.nlm.nih.gov/nuccore/5420376, https://www.ncbi.nlm.nih.gov/nuccore/13122261, https://www.ncbi.nlm.nih.gov/nuccore/386646758, https://www.ncbi.nlm.nih.gov/nuccore/295311559, 2017-01-24 09:40:17, 2017-01-24 09:40:17, 2017-01-24 09:40:17, 2017-01-24 09:40:17, 2017-01-24 09:40:17
#&gt;                                                                                                                                         input_list
#&gt; 1 https://example.com/dna.cgi?cmd=objFile&amp;ids=514683, https://example.com/dna.cgi?cmd=objFile&amp;ids=514682, 2017-01-24 09:40:17, 2017-01-24 09:40:17
#&gt;                                                                                                                                       output_list
#&gt; 1 https://example.com/data/514769/allCount-aligned.csv, https://example.com/data/514801/SNPProfile*.csv, 2017-01-24 09:40:17, 2017-01-24 09:40:17
#&gt; 
#&gt; 
#&gt; $execution_domain
#&gt; $execution_domain$script
#&gt; [1] "https://example.com/workflows/antiviral_resistance_detection_hive.py"
#&gt; 
#&gt; $execution_domain$script_driver
#&gt; [1] "shell"
#&gt; 
#&gt; $execution_domain$software_prerequisites
#&gt;            name      version
#&gt; 1  HIVE-hexagon babajanian.1
#&gt; 2 HIVE-heptagon   albinoni.2
#&gt;                                                                                                                    uri
#&gt; 1 https://example.com/dna.cgi?cmd=dna-hexagon&amp;cmdMode=-, 2017-01-24 09:40:17, d60f506cddac09e9e816531e7905ca1ca6641e3c
#&gt; 2                                      https://example.com/dna.cgi?cmd=dna-heptagon&amp;cmdMode=-, 2017-01-24 09:40:17, NA
#&gt; 
#&gt; $execution_domain$external_data_endpoints
#&gt; $execution_domain$external_data_endpoints[[1]]
#&gt;           name                                     url
#&gt; 1 generic name protocol://domain:port/application/path
#&gt; 
#&gt; $execution_domain$external_data_endpoints[[2]]
#&gt;                   name                        url
#&gt; 1 access to ftp server ftp://data.example.com:21/
#&gt; 
#&gt; $execution_domain$external_data_endpoints[[3]]
#&gt;                            name                                           url
#&gt; 1 access to e-utils web service https://eutils.ncbi.nlm.nih.gov/entrez/eutils
#&gt; 
#&gt; 
#&gt; $execution_domain$environment_variables
#&gt; $execution_domain$environment_variables$HOSTTYPE
#&gt; [1] "x86_64-linux"
#&gt; 
#&gt; $execution_domain$environment_variables$EDITOR
#&gt; [1] "vim"
#&gt; 
#&gt; 
#&gt; 
#&gt; $parametric_domain
#&gt;                          param value step
#&gt; 1                         seed    14    1
#&gt; 2            minimum_match_len    66    1
#&gt; 3 divergence_threshold_percent  0.30    1
#&gt; 4             minimum_coverage    15    2
#&gt; 5                  freq_cutoff  0.10    2
#&gt; 
#&gt; $io_domain
#&gt; $io_domain$input_subdomain
#&gt;                                                                                                                 uri
#&gt; 1             Hepatitis C virus genotype 1, https://www.ncbi.nlm.nih.gov/nuccore/22129792, 2017-01-24T09:40:17-0500
#&gt; 2 Hepatitis C virus type 1b complete genome, https://www.ncbi.nlm.nih.gov/nuccore/5420376, 2017-01-24T09:40:17-0500
#&gt; 
#&gt; $io_domain$output_subdomain
#&gt;   mediatype
#&gt; 1  text/csv
#&gt; 2  text/csv
#&gt;                                                                               uri
#&gt; 1 https://example.com/data/514769/dnaAccessionBased.csv, 2017-01-24T09:40:17-0500
#&gt; 2       https://example.com/data/514801/SNPProfile*.csv, 2017-01-24T09:40:17-0500
#&gt; 
#&gt; 
#&gt; $error_domain
#&gt; $error_domain$empirical_error
#&gt; $error_domain$empirical_error$false_negative_alignment_hits
#&gt; [1] "&lt;0.0010"
#&gt; 
#&gt; $error_domain$empirical_error$false_discovery
#&gt; [1] "&lt;0.05"
#&gt; 
#&gt; 
#&gt; $error_domain$algorithmic_error
#&gt; $error_domain$algorithmic_error$false_positive_mutation_calls
#&gt; [1] "&lt;0.00005"
#&gt; 
#&gt; $error_domain$algorithmic_error$false_discovery
#&gt; [1] "0.005"
#&gt; 
#&gt; 
#&gt; 
#&gt; attr(,"class")
#&gt; [1] "list" "bco" </div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Jeffrey Grover, Soner Koc, Nan Xiao, Dennis Dean, Seven Bridges Genomics.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


